src/NexusMods.CLI/CliGuidedInstaller.cs

```diff
@@ -0,0 +1,241 @@
+    private static readonly object[] TableOfOptionsFooterBackToGroupSelection = { BackInput, "", "Back", "Back to the group selection" };
+    private static readonly object[] TableOfOptionsFooterCancel = { CancelInput, "", "Cancel", "Cancel the installation" };
+
+    private readonly ILogger<CliGuidedInstaller> _logger;
+
+    /// <summary>
+    /// Constructor.
+    /// </summary>
+    public CliGuidedInstaller(ILogger<CliGuidedInstaller> logger)
+    {
+        _logger = logger;
+    }
+
+    /// <summary>
+    /// The renderer to use for rendering the options.
+    /// </summary>
+    public IRenderer Renderer { get; set; } = null!;
+
+    /// <summary>
+    /// If true, all option selection will select the first choice.
+    /// </summary>
+    public bool SkipAll { get; set; }
+
+    /// <inheritdoc/>
+    public void SetupInstaller(string name) { }
+
+    /// <inheritdoc/>
+    public void CleanupInstaller() { }
+
+    /// <inheritdoc />
+    public Task<UserChoice> RequestUserChoice(GuidedInstallationStep installationStep, CancellationToken cancellationToken)
+    {
+        OptionGroup? currentGroup = null;
+
+        var selectedOptions = installationStep.Groups
+            .SelectMany(group => group.Options
+                .Where(option => option.Type == OptionType.PreSelected)
+                .Select(option => new SelectedOption(group.Id, option.Id))
+            ).ToList();
+
+        while (true)
+        {
+            if (currentGroup is null)
+            {
+                // if the installation step has multiple groups
+                // the user has to select which group they want to use
+
+                RenderTableOfGroups(installationStep);
+
+                var input = SkipAll ? NextInput : GetUserInput();
+
+                switch (input)
+                {
+                    case CancelInput:
+                        return Task.FromResult(new UserChoice(new UserChoice.CancelInstallation()));
+                    case BackInput:
+                        return Task.FromResult(new UserChoice(new UserChoice.GoToPreviousStep()));
+                    case NextInput:
+                    {
+                        var requiredOptions = installationStep.Groups
+                            .SelectMany(group => group.Options
+                                .Where(option => option.Type == OptionType.Required)
+                                .Select(option => new SelectedOption(group.Id, option.Id))
+                            );
+
+                        selectedOptions.AddRange(requiredOptions);
+
+                        // proceed to the next step
+                        return Task.FromResult(new UserChoice(new UserChoice.GoToNextStep(selectedOptions.ToArray())));
+                    }
+                }
+
+                var groupIndex = ParseNumericalUserInput(input, installationStep.Groups.Length);
+                if (groupIndex < 0) continue;
```

Comment:
For better code clarity, shouldn't the following code be part of the switch? Either as default case or I think C# has some cool tricks with switch to allow handling the parse output and the index out of ranges as separate cases, but just the default would already be clearer I think.